### PR TITLE
Add nif_image_write_to_array to copy the matrix

### PR DESCRIPTION
**Pull request by kipcole9**
Adds nif_image_write_to_array to copy the matrix
of any image to be able to interop with other
libraries like eVision and Nx.

Adds Vix.Vips.Image.write_to_array/1 for the
public API to return a reference to the image matrix
that can be passed to other Elixir/erlang libraries.

Not ready for production use but would welcome your
input and guidance to see if I'm heading the right way.

Next step will be the reverse: `new_from_array` to consume
arrays from other libraries.

In this version the matrix is copied. This to avoid memory
ownership issues. It would be desirable to find a memory
safe way to share the image matrix but that comes with both
concurrency and memory safe issues.

### CODE REVIEW

**Comment by kipcole9** [comment]
Excited to see you merge the pipe source/sink branch into master as well!

File: c_src/vips_image.c [hunk 1]
```diff
@@ -220,6 +220,40 @@ ERL_NIF_TERM nif_image_write_to_buffer(ErlNifEnv *env, int argc,
...
+
+  if (!array) {
+    error("Failed to write VipsImage to array. error: %s",
+          vips_error_buffer());
+    vips_error_clear();
+    ret = make_error(env, "Failed to write VipsImage to array");
+    goto exit;
+  }
+
+  ret = make_ok(env, g_object_to_erl_term(env, (GObject *)array));
```
Comment (akash-akya) [inline c_src/vips_image.c:250]: `array` in this case it is a pointer to a raw memory block, so we can't treat it like [g_object](https://docs.gtk.org/gobject/concepts.html#the-gobject-base-class). 

There are a couple of options here, we can convert it to binary and transfer the ownership, or we can use `GBoxed` (or libvips [abstraction](https://www.libvips.org/API/current/libvips-type.html#VipsArea) over it), or build our own solution. We also need to consider how are we going to support concurrency with multiple schedulers accessing the same memory.

**Review by akash-akya — COMMENTED** [review]
@kipcole9 thanks for the MR.

Agree with you that we should start focusing on the interop aspect. Unfortunately, I have limited knowledge of Nx & eVision and targeted access patterns/use cases. I'll try to find some time to understand this better and get back to you on this :)

**Comment by kipcole9** [comment]
Thanks for the feedback and yes, there are definitely challenges in sharing access to the underlying data safely. I hope you will have a flash of inspiration sometime!

In the meantime, I have reverted `nif_image_write_to_array` and implemented the very similar `nit_image_write_to_binary`  to take the image data, create a binary copy, and return that.  A lot less efficient but a lot safer.

There are some open items for your consideration:
1. I am using a copy of the `Nx.Tensor.t` type to package the binary, the tensor shape and the format. I think using this format will aid other integrations with other you and other library builders are willing.
2. I don't have a good way to translate the `Vix.Vips.Enun.VipsBandFormat.t` types into native data types - I don't there is currently a Vips.Nif function to return the native data type sizes?
3.

**Comment by kipcole9** [comment]
With your zero-copy code in place I think we open up a lot of opportunities (and its how PyVips integrates with Numpy and PIL as well).  Hopefully eVision will follow along.

As for the how do we share meta data about a memory blob, I think having a canonical form helps. I copied the Nx tensor struct approach not to tie closely together but more to advocate for a consistent strategy for sharing amongst libraries and I think the Nx struct isn't a bad approach.  What do you think?

I don't believe there is an API in the BEAM to return the size of native types but you will know that much better than I do.  If there isn't one then I think even I can create a NIF function to return that data. Let me know?

File: lib/vix/tensor.ex [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  defp nx_type(image) do
+    word_size = :erlang.system_info(:wordsize)
+
+    case Image.format(image) do
+      :VIPS_FORMAT_UCHAR   -> {:u, 8}
+      :VIPS_FORMAT_CHAR    -> {:s, 8}
+      :VIPS_FORMAT_USHORT  -> {:u, 16}
+      :VIPS_FORMAT_SHORT   -> {:s, 16}
+      :VIPS_FORMAT_UINT    -> {:u, word_size}
```
Comment (akash-akya) [inline lib/vix/tensor.ex:36]: I think in libvips bitsize is fixed, and does not depend on the machine architecture. 

* https://github.com/libvips/pyvips/blob/df462ca09b8d63c8a5a9cd2f21a6564ddcd3ef08/pyvips/vimage.py#L131-L141
* https://github.com/libvips/libvips/blob/6260a37136973f3d231d0e2c28af6190126c664e/libvips/foreign/matlab.c#L147-L156

wdyt?

**Review by akash-akya — COMMENTED** [review]
Sorry for the slow response from my end. 
I got caught up in other things.

File: lib/vix/tensor.ex [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    names: [:x, :y, :band],
+    type: {}
+  ]
+
+  def binary_to_tensor(binary, size, %Image{} = image)
+      when is_binary(binary) and is_integer(size) do
+    struct(__MODULE__,
+      data: binary,
+      shape: {Image.width(image), Image.height(image), Image.bands(image)},
+      names: [:x, :y, :band],
```
[the remark this answers is not in the document]
Comment (kipcole9) [reply lib/vix/tensor.ex:18]: Makes perfect sense, I will update the PR. I've also got a pretty intense work week this week so probably not until the weekend.

File: lib/vix/tensor.ex (hunk 2 above)
Comment (kipcole9) [reply lib/vix/tensor.ex:36]: I think that's really great :-)

File: lib/vix/vips/image.ex [hunk 4]
```diff
@@ -305,6 +305,38 @@ defmodule Vix.Vips.Image do
...
+  In VIPS, images are uninterpreted arrays, meaning that from
+  the point of view of most operations, they are just large
+  collections of numbers. There's no difference between an RGBA
+  (RGB with alpha) image and a CMYK image, for example, they are
+  both just four-band images.
+
+  This function is intended to support interoperability of image
+  data between different libraries.  Since the array is created as
+  a NIF resource it will be correctly garbage collected when
+  the last reference falls out of scope.
```
[the remark this answers is not in the document]
Comment (kipcole9) [reply lib/vix/vips/image.ex:328]: Thanks of the comment. I think good documentation is often what makes the difference between a good library, and a good library that gets used :-)

File: lib/vix/tensor.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (kipcole9) [reply lib/vix/tensor.ex]: Updated to name the struct fields `:width`, `:height` and `:bands`.

File: lib/vix/tensor.ex (hunk 2 above)
Comment (akash-akya) [reply lib/vix/tensor.ex:36]: So shall we change this to use `32` and `64` here?

**Pull request merged**